### PR TITLE
HTTP CloudEvent coercion

### PR DESCRIPTION
**Pull request by raeesbhatti**
Add supports for proper coercion of HTTP and Invoke events to CloudEvent format.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [ ] Provide verification commands / resources
- [ ] Enable "Allow edits from maintainers" for this PR
- [ ] Update the messages below

***Is this ready for review?:*** NO
***Is it a breaking change?:*** NO

### CODE REVIEW

File: event/request.go [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+
+func extractEventType(r *http.Request) Type {
+	eventType := Type(r.Header.Get("event"))
+	if eventType == "" {
+		eventType = TypeHTTP
+	}
+	return eventType
+}
+
+func parseAsCloudEventBinary(event *Event, headers http.Header) *Event {
```
Comment (mthenw) [inline event/request.go]: One more request here. It should not get Event as argument. It should be similar to `parseAsCloudEvent`. So the arguments should be only what's needed which is headers and `body []byte`. Event should be created inside.

File: event/http.go [hunk 2]
```diff
@@ -18,10 +18,10 @@ type HTTPEvent struct {
 }
 
 // NewHTTPEvent returns a new instance of HTTPEvent
-func NewHTTPEvent(r *http.Request, eventData interface{}) *HTTPEvent {
+func NewHTTPEvent(r *http.Request, eventData interface{}) *HTTPRequestData {
```
Comment (mthenw) [inline event/http.go]: This should also be renamed to NewHTTPRequestData

File: event/request.go [hunk 3]
```diff
@@ -0,0 +1,96 @@
+package event
```
Comment (mthenw) [inline event/request.go]: can you move the content of this file to event.go? those file are pretty short and the are basically about the same thing.

File: event/event_test.go [hunk 4]
```diff
@@ -60,141 +101,242 @@ var newTests = []struct {
...
+	{
+		[]byte(`{"hello": "world"}`),
 		"application/json",
-		[]byte(`{
+		map[string]interface{}{"hello": "world"},
+	},
+}
+
+var fromRequestTests = []struct {
+	url           string
```
Comment (mthenw) [inline event/event_test.go]: url doesn't matter here, we can remove it.

File: event/event.go [hunk 5]
```diff
@@ -116,13 +149,55 @@ func (e Event) MarshalLogObject(enc zapcore.ObjectEncoder) error {
...
 
+func extractEventType(r *http.Request) Type {
+	eventType := Type(r.Header.Get("event"))
+	if eventType == "" {
+		eventType = TypeHTTP
+	}
+	return eventType
+}
+
+func parseAsCloudEventBinary(event *Event, headers http.Header) *Event {
```
Comment (mthenw) [inline event/event.go]: I mentioned that in the previous review. It should not get Event as an argument. It should be similar to `parseAsCloudEvent`. So the arguments should be only what's needed which is headers and `body []byte`. Event should be created inside.

File: event/event_test.go (hunk 4 above)
Comment (raeesbhatti) [reply event/event_test.go]: Well, it actually does. Because `NewHTTPRequestData` expects a URL.

Comment (raeesbhatti) [reply event/event_test.go]: In the `http.Request` that we're passing to `FromRequest` function

File: event/event.go [hunk 6]
```diff
@@ -116,40 +154,103 @@ func (e Event) MarshalLogObject(enc zapcore.ObjectEncoder) error {
...
+		event.SchemaURL = val
+	}
+
+	event.Extensions = map[string]interface{}{}
+	for key, val := range ihttp.FlattenHeader(headers) {
+		if strings.HasPrefix(key, "Ce-X-") {
+			key = strings.TrimLeft(key, "Ce-X-")
+			// Make first character lowercase
+			runes := []rune(key)
+			runes[0] = unicode.ToLower(runes[0])
```
[the remark this answers is not in the document]
Comment (mthenw) [reply event/event.go]: yeah, that's true, but in the same time, when you look at the [example in HTTP binding spec](https://github.com/cloudevents/spec/blob/master/http-transport-binding.md#3131-http-header-names) you will see that desired behavior is to preserve case (`testExtension`).

**Pull request merged**